### DISCUSSION TITLE
Ability to Instrument Existing Classes

### DISCUSSION THREAD

**Discussion by sirinath** [activej/activej#183] [Ideas]
Can we have the ability to instrument existing classes? e.g.

```java
rebase("AClass")
.field("a").initilise(instentiate("MyObj").param(1).at(0).of(int.class))
.method("f").body(statements(call(self().field("a").method("x").params(args(0))), origin().method("f").params(args(0)))))
```

Introduce a field `a` and initialise it with constructor taking an `int` and call its method `x` in method `f` followed by the original implementation of `f`.

Instrumentation can happen at:
- selected or all classes at runtime
- selected or all classes at load time
- selected or all classes in a Jar file
- selected or all before Jar file packing

To find what classes to instrument one can use something like [ClassGraph](https://github.com/classgraph/classgraph)

**Comment by eduard-vasinskyi** [comment]
Hello, @sirinath 
We are not planning to add class instrumentation to ActiveJ. 
I believe you can use Java’s Instrumentation API with or without  ActiveCodegen for this purpose. If you have trouble with this approach or if you mean something else, please let us know!

**Comment by sirinath** [comment]
Using 2 different bytecode manipulation libraries are not convenient. So for ActiveCodegen to be useful it may need to gain some parity with other frameworks out there like:
- https://bytebuddy.net/#/
- https://www.javassist.org/
- https://github.com/Guardsquare/proguard-core
- etc.

As for the use cases I have are:
- adding removing loggers and benchmark code into functions, try catch blocks, loops, etc.
- replacing methods with new implementations
- etc.

**Comment by eduard-vasinskyi** [comment]
ActiveCodegen was designed for a completely different purpose - as a tool for generating new classes. If you need to transform classes, you should probably use a library suited for that task. We currently do not see a need to add class instrumentation to ActiveCodegen. Or, alternatively, you can use ActiveCodegen to generate subclasses for specified classes - if that suits your goals, of course.
